### ISSUE TITLE
False 404 index_not_found_exception from /_cat/indices/ALIAS after master returns

### ISSUE THREAD

**Issue by jamshid** [elastic/elasticsearch#70572]
**Elasticsearch version** (`bin/elasticsearch --version`):
```
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Version: 7.5.2, Build: default/rpm/8bec50e1e0ad29dad5653712cf3bb580cd1afcdf/2020-01-15T12:11:52.313576Z, JVM: 13.0.1
**Plugins installed**: []
```
**JVM version** (`java -version`):
```
openjdk version "13.0.1" 2019-10-15
OpenJDK Runtime Environment AdoptOpenJDK (build 13.0.1+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 13.0.1+9, mixed mode, sharing)
```
**OS version** (`uname -a` if on a Unix-like system):
CentOS Linux release 7.9.2009 (Core)

**Description of the problem including expected versus actual behavior**:

After master node has been down, when it returns there is a brief window where `/_cat/indices/ALIAS` requests are accepted but elasticsearch returns a 404 `index_not_found_exception`. This is unexpected and problematic behavior to our app which considers this an error.

I'd expect a usually strongly consistent system like elasticsearch to return a 503 if it hasn't fully read enough data after master returned to determine if the alias+index exists. It's ok to give an error, it's not okay to give false information like that the index does not exist.

**Steps to reproduce**:
Please include a *minimal* but *complete* recreation of the problem,
including (e.g.) index creation, mappings, settings, query etc.  The easier
you make for us to reproduce it, the more likely that somebody will take the
time to look at it.

 1. For simplicity create a 3 node cluster with one node being the only master and not accepting data.
 2. Create an index with an alias MYALIAS.
 3. Run a curl loop below requesting an index by its alias name to a data node, verify it returns 200.
 4. Stop the master node, the curls soon fail to connect `curl: (7) Failed to connect`.
 5.  Bring the master node back up, it starts  returning `curl: (52) Empty reply from server`
 6. I see one false 404 returned before it goes back to returning a 200 (with a briefly red index).
```
$ while sleep 1 ; do curl -i ELASTICSEARCH:9200/_cat/indices/MYALIAS  ; done
...
HTTP/1.1 404 Not Found
content-type: application/json; charset=UTF-8
content-length: 437

{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [MYALIAS]","resource.type":"index_or_alias","resource.id":"MYALIAS","index_uuid":"_na_","index":"MYALIAS"}],"type":"index_not_found_exception","reason":"no such index [MYALIAS]","resource.type":"index_or_alias","resource.id":"MYALIAS","index_uuid":"_na_","index":"MYALIAS"},"status":404}
```

**Provide logs (if relevant)**:

Sorry I can try to recover logs if you can't reproduce this.

**Comment by fcofdez** [comment]
I've been trying to reproduce the issue that you're describing but I got the expected results.

```yaml
version: '3.7'

services:
  es01:
    image: "docker.elastic.co/elasticsearch/elasticsearch-oss:7.5.2"
    environment:
      node.name: es01
      discovery.seed_hosts: es02
      cluster.initial_master_nodes: es01
      cluster.name: test
      bootstrap.memory_lock: "true"
      node.master: "true"
      node.data: "false"
      ES_JAVA_OPTS: -Xms256m -Xmx256m
    ports:
      - "9200:9200"
    volumes:
      - "es-data-es01:/usr/share/elasticsearch/data"
    ulimits:
      memlock:
        soft: -1
        hard: -1
  es02:
    image: "docker.elastic.co/elasticsearch/elasticsearch-oss:7.5.2"
    environment:
      node.name: es02
      discovery.seed_hosts: es01
      cluster.initial_master_nodes: es01
      cluster.name: test
      bootstrap.memory_lock: "true"
      node.master: "false"
      node.data: "true"
      ES_JAVA_OPTS: -Xms256m -Xmx256m
    volumes:
      - "es-data-es02:/usr/share/elasticsearch/data"
    ulimits:
      memlock:
        soft: -1
        hard: -1
    ports:
      - "9202:9200"
  es03:
    image: "docker.elastic.co/elasticsearch/elasticsearch-oss:7.5.2"
    environment:
      node.name: es03
      discovery.seed_hosts: es01
      cluster.initial_master_nodes: es01
      cluster.name: test
      bootstrap.memory_lock: "true"
      node.master: "false"
      node.data: "true"
      ES_JAVA_OPTS: -Xms256m -Xmx256m
    volumes:
      - "es-data-es03:/usr/share/elasticsearch/data"
    ulimits:
      memlock:
        soft: -1
        hard: -1
    ports:
      - "9203:9200"

volumes:
  es-data-es01:
  es-data-es02:
  es-data-es03:
```

With this setup and requesting an alias to the master node I cannot see any `404` and if I query through a data node I get the expected `503` status code with `master_not_discovered_exception` if the request times out (when the master node is down). Could you share more details about the configuration where you experience that behaviour @jamshid? Thanks!

**Comment by jamshid** [comment]
At first I couldn't reproduce it in that docker compose environment either but then I added more indices/data.

Try creating 1000 indices:

```
# Create a test index and alias
$ curl -i -XPUT http://localhost:9200/test
$ curl -i -XPUT http://localhost:9200/test/_alias/testalias

# Create a 1000 indices
$ seq 1000 | xargs -t -P 10 -n 1 -I{} curl -fsS -XPUT "http://localhost:9200/test-{}"
```
Monitor this loop while running `docker-compose stop es01 ; sleep 10 ; docker-compose start es01` and you should see a 404.

```
$ while sleep 1 ; do curl -ifsS -m 5  "http://localhost:920[2-3]/_cat/indices/testalias"  ; done
...
HTTP/1.1 200 OK
content-type: text/plain; charset=UTF-8
content-length: 62

green open test 3eYTLaUuRYOsq_osjBgX4g 1 1 159 0 17.5kb 8.7kb
curl: (22) The requested URL returned error: 429 Too Many Requests
HTTP/1.1 200 OK
content-type: text/plain; charset=UTF-8
content-length: 61

green open test 3eYTLaUuRYOsq_osjBgX4g 1 1 159 0 8.7kb 8.7kb
curl: (28) Operation timed out after 5001 milliseconds with 0 bytes received
curl: (28) Operation timed out after 5000 milliseconds with 0 bytes received
curl: (28) Operation timed out after 5000 milliseconds with 0 bytes received
curl: (28) Operation timed out after 5000 milliseconds with 0 bytes received
curl: (22) The requested URL returned error: 429 Too Many Requests
curl: (22) The requested URL returned error: 429 Too Many Requests
curl: (22) The requested URL returned error: 429 Too Many Requests
curl: (22) The requested URL returned error: 429 Too Many Requests
curl: (22) The requested URL returned error: 429 Too Many Requests

curl: (28) Operation timed out after 5004 milliseconds with 0 bytes received
curl: (28) Operation timed out after 5000 milliseconds with 0 bytes received
curl: (28) Operation timed out after 5005 milliseconds with 0 bytes received
curl: (28) Operation timed out after 5000 milliseconds with 0 bytes received
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
HTTP/1.1 200 OK
content-type: text/plain; charset=UTF-8
content-length: 45

red open test 3eYTLaUuRYOsq_osjBgX4g 1 1    
HTTP/1.1 200 OK
content-type: text/plain; charset=UTF-8
content-length: 45

red open test 3eYTLaUuRYOsq_osjBgX4g 1 1    
HTTP/1.1 200 OK
content-type: text/plain; charset=UTF-8
content-length: 45

red open test 3eYTLaUuRYOsq_osjBgX4g 1 1    
```

**Comment by fcofdez** [comment]
@jamshid thanks for the detailed reproduction scenario. I've been able to reproduce and find where this error is coming from. I'll open a PR with a fix.

**Comment by jamshid** [comment]
Thanks @fcofdez! Is there a way to see which release this will be in, hopefully the next 7.x?

**Comment by fcofdez** [comment]
Thanks for your thorough reproduction scenario @jamshid! It will be released in 7.16 👍
